R/callback-validation-check.R

```diff
@@ -0,0 +1,74 @@
+#' @include callbacks.R
+NULL
+
+#' Validation Check
+#'
+#' Check validation loop before fitting model.
+#'
+#' @param batches Number of validation batches to check. Default is 2.
+#'
+#' @note Usually the training loop is much longer than the validation
+#'   loop and issues with the validation loop aren't encountered until after
+#'   a long training runtime. This callback runs the validation loop first on
+#'   `batches` number of batches and then proceeds onto the standard
+#'   training process.
+#'
+#' @note Printing can be disabled by passing `verbose = FALSE` to
+#'   [fit.luz_module_generator()].
+#'
+#' @family luz_callbacks
+#'
+#' @returns
+#' A `luz_callback`.
+#'
+#' @export
+luz_callback_validation_check <- luz_callback(
+  "validation_check_callback",
+  initialize = function(batches = 2) {
+    if (!rlang::is_scalar_integerish(batches)) {
+      rlang::abort("`batches` must be a single integer value.")
+    }
+    self$batches <- batches
+  },
+  on_fit_begin = function() {
+    if (is.null(ctx$valid_data)) return()
+    if (self$batches <= 0) return()
+
+    ctx$model$eval()
+    ctx$training <- FALSE
+
+    self$initialize_progress_bar()
+
+    i <- 0
+    torch::with_no_grad({
+      coro::loop(for (batch in ctx$valid_data) {
+        self$validate_one_batch(batch)
+        self$tick_progress_bar(self$loss)
+        i <- i + 1
+        if (i >= self$batches) break()
+      })
+    })
+  },
+  validate_one_batch = function(batch) {
+    input  <- list(batch[[1]])
+    target <- batch[[2]]
+    pred <- do.call(ctx$model, input)
+    self$loss <- ctx$model$loss(pred, target)
```

Comment:
I think in general we would want to do the full validation step because the errors could be in any of the callbacks etc. But we would need to take care of the side effects that this might cause.

We would need to call `valid_one_step()` and then make sure we can reset the state. Not sure yet what would be the best way to do it though.

Reply:
I was thinking about this problem. It feels to me that the safest way would be to `on_fit_begin()` we call `fit` again and we add a callback that breaks the training loop after `batches` steps for both training and validation. This way, no side effects would interfere in the actual training loop but we still run the full loop which would detect the other possible bugs.

I think this is possible if the first thing we do in the `ctx` object is to save a list with all arguments that were passed to 
`fit`, before we do any kind of manipulation (like we do for callbacks). 
To avoid the infinite recursion we could check `ctx$callbacks` to check if the callback that breaks the loop is present.

Reply:
Yeah I've kind of gone in circles here. We want to call the validation callbacks so it is a complete check of the validation loop, but I was worried about any changes in state this might have. I did consider using `valid_one_batch()` but at the time decided against it for the above reasons.

Reply:
Somewhat related question, when `ctx$call_callbacks("on_...")` is called, if there are multiple callbacks for a breakpoint, what is the order they are called in? Default callbacks first, user-supplied callbacks second?

Reply:
Yes, they are called in that order: default callbacks then user callbacks.
I think that if we call `fit` again, there would be no interference, the only difference is that it would also test the training loop. But we could also skip it anyway...

Reply:
I did actually think about calling `fit()` again inside `on_fit_begin()` but I decided against. But you're right, it would be a good way to check both the training and validation loops before committing to a full fit.

Reply:
Thinking again, there could still be some side effects, eg: the callbacks passed by the user can have side effects outside of the R session (maybe writing to a file or something like this). So maybe we want to call fit again, only with the default callbacks + the one that breaks the training loop. 

This is not completely ideal, because still there would be callbacks that could fail in the 'real' pass.

Reply:
I agree with only calling the default callbacks. My original reason for avoiding callbacks was for loggers and other things written to disk. But if we only run default callbacks we can avoid this issue. The function docs can just point out that user callbacks aren't validated.